Sidebar updates

mortenpi opened the pull request:
Style and behaviour updates for the sidebar. Site build [available here](http://mortenpi.eu/Documenter.jl/fix-style/).

- Improve contrast of the sidebar menu as discussed [here](https://discourse.julialang.org/t/improve-contrast-in-docs-julialang-org-menu/1370/11).
- Update the version selector style and behaviour. Basically to make fit a bit better with the rest of the style and also make sure that it displays the current version when loaded.

The WIP bit is to make `makedocs` automatically determine the current version name (e.g. `latest`, tag name) which is currently only done in `deploydocs` (so, right now it would just show the commit hash for all builds).

I am also not necessarily entirely happy with the version selector's style, so feedback & ideas would be greatly appreciated (cc @cormullion).

Ideally fixes #406.

cormullion commented:
Looks pretty good—works well on macOS/Safari. I might prefer it if the dropdown menu and default text in the search box (and the logo) was all centered... 

On iPad (and probably iPhone too but...) the navigation  bar isn't very usable, though —touch event handling or something. Perhaps that could be added as a separate issue...

MichaelHatherly commented:
> Update the version selector style and behaviour.

The selector doesn't display the dropdown arrow as far as see. Could we add that back since it's a useful indication of what it is.

> The WIP bit is to make `makedocs` automatically determine the current version name

`makedocs` isn't currently dependant on the "version" at all. Changing that would probably require merging `makedocs` and `deploydocs`. Best way to handle this in `deploydocs` would be to create a `version.js` file containing `var DOCUMENTER_VERSION = "<dirname>";` for each version directory. Then use that value for the selector.

mortenpi commented:
> The selector doesn't display the dropdown arrow as far as see. Could we add that back since it's a useful indication of what it is.

Yup, I agree that it would be nice to have. However, I removed it intentionally since it can't be styled and it is very inconsistent across platforms/browsers and doesn't really match with the rest of the style (especially in Firefox). But I should play around with it a bit more -- I think the standard workaround is to use background images on the `select` to show an arrow.

> Best way to handle this in `deploydocs` would be to create a `version.js` file containing `var DOCUMENTER_VERSION = "<dirname>";` for each version directory. Then use that value for the selector.

It would be nice if it wouldn't depend on javascript though and the current version would be displayed statically. But this sounds like a reasonable solution for the first iteration.

mortenpi commented:
> I might prefer it if the dropdown menu and default text in the search box (and the logo) was all centered... 

Everything except the text in the search box should be centered. Could you give me a screenshot of what you're seeing?

> On iPad (and probably iPhone too but...) the navigation  bar isn't very usable, though —touch event handling or something. Perhaps that could be added as a separate issue...

Mobile experience from both the layout and touch experience point of view has not received any attention so far really. Would definitely be something that could be improved.

MichaelHatherly commented:
> especially in Firefox

Looks alright for me on Firefox, I think it depends on what dm/wm and theme is being used.

![2017-01-19-1484808444-crop](https://cloud.githubusercontent.com/assets/6144086/22096443/369966de-de24-11e6-8116-c9190d567ca9.png)

> It would be nice if it wouldn't depend on javascript

There's already lots that depends on js in the docs, so for a first pass it's fine to require it for this.

> Mobile experience from both the layout and touch experience point of view has not received any attention so far really. Would definitely be something that could be improved.

According to the site analytics mobile views account for less than 5% of all traffic to Documenter's docs, so probably not a critical requirement just yet.

cormullion commented:
On Safari I see this:

![screen shot 2017-01-19 at 10 05 46](https://cloud.githubusercontent.com/assets/52289/22102596/e6d11078-de2f-11e6-8fe4-ad0013edbf37.png)

Artist's impression of slight improvement (_de gustibus non est disputandum_ ...):

![screen shot 2017-01-19 at 10 05 46 copy copy](https://cloud.githubusercontent.com/assets/52289/22102649/186c5cbe-de30-11e6-88c6-ad09a2974727.png)

mortenpi commented:
For me on Firefox, when I re-enable the arrow, it looks like this:
![documenter_ff](https://cloud.githubusercontent.com/assets/147757/22112067/54622a5c-de6a-11e6-922b-98e1560957a7.png)

However, I agree that there should be something indicating that you can actually select a version there, so I'll look into that. Also, agreed, let's go with a `.js` file for the time being.

As for centering -- the `select` has `text-align: center` and `text-align-last: center`, which means it is nicely centered on all the browsers that I've tested (but I don't have access to OSX/Safari). What styling did you add, @cormullion?

The search box I would keep left-aligned since it's quite wide and then it matched the menu below.

Just FYI: I'll be travelling in the coming days though, so this PR will likely be on hold until the latter half of next week.

MichaelHatherly commented:
Looks like a 0.4-specific failure. You're welcome to drop 0.4 support if you'd like to &mdash; we've been keeping it going for long enough I think.

mortenpi commented:
It's a small type thing with `String`s -- as a workaround I think can force the string to be `Compat.UTF8String` so it works on all versions.

mortenpi commented:
Updated on my part, latest build should be [here](http://mortenpi.eu/Documenter.jl/latest/) (built and deployed on [Travis](https://travis-ci.org/mortenpi/Documenter.jl/jobs/195858084)). Changes:

* Unified select box triangle with a background image.
* Uses `siteinfo.js` to populate the selector's "current version".

I do feel that this needs more work in the long run. Generating `siteinfo.js` in `deploydocs` doesn't feel quite right, and the `latest`/`release` etc. tags should probably have the exact version shown as well (e.g. "v0.5.0 (latest)" instead of just "latest"). But all that should be in future PRs.

mortenpi commented:
Cleaned up and rebased, [latest build](http://mortenpi.eu/Documenter.jl/latest/) updated. I would say that it's ready to merge.

Only thing unaddressed is the Safari style, which I can't test myself. @cormullion ?

cormullion commented:
Looks good to me!

![img_0074](https://cloud.githubusercontent.com/assets/52289/22625080/1ba439f4-eb84-11e6-92ad-02fbb12373d2.PNG)

mortenpi commented:
The text in the selector should be centered (and is on other platforms). What did you add before to make it centered in Safari, @cormullion?

cormullion commented:
My artist's impression above, you mean? That was image editing... I moved the stuff around manually. That was after a quick look at the Safari Inspector thing, which lets you fiddle with the HTML/CSS. (I think Safari on iPad and Mac will be the same. I'll have a look after breakfast...☕️🥐)

cormullion commented:
Safari on Mac:

![screen shot 2017-02-05 at 10 24 05](https://cloud.githubusercontent.com/assets/52289/22625404/461b4930-eb8d-11e6-83ae-b7640c0cf1fd.png)

mortenpi commented:
Ah, I thought you did it in the inspector. I added a note about this for future reference as it looks OK even if it is not centered.
